docs/advanced-features/middleware.md

```diff
@@ -1,95 +1,166 @@
 ---
-description: Learn how to use Middleware in Next.js to run code before a request is completed.
+description: Learn how to use Middleware to run code before a request is completed.
 ---
 
-# Middleware (Beta)
+# Middleware
 
 <details open>
-  <summary><b>Version History</b></summary>
+<summary><b>Version History</b></summary>
 
-| Version   | Changes                                                                                                 |
-| --------- | ------------------------------------------------------------------------------------------------------- |
-| `canary`  | Preparing for stability, see [upgrade guide](https://nextjs.org/docs/messages/middleware-upgrade-guide) |
-| `v12.0.9` | Enforce absolute URLs in Edge Runtime ([PR](https://github.com/vercel/next.js/pull/33410))              |
-| `v12.0.0` | Middleware (Beta) added.                                                                                |
+| Version | Changes |
+
+| --------- | ------------------------------------------------------------------------------------------ |
+
+| `v12.2.0` | Middleware is stable |
+
+| `v12.0.9` | Enforce absolute URLs in Edge Runtime ([PR](https://github.com/vercel/next.js/pull/33410)) |
+
+| `v12.0.0` | Middleware (Beta) added |
 
 </details>
 
-Middleware enables you to run code before a request is completed. Based on the user's incoming request, you can modify the response by rewriting, redirecting, adding headers, or even streaming HTML.
+Middleware allows you to run code before a request is completed, then based on the incoming request, you can modify the response by rewriting, redirecting, adding headers, or setting cookies.
 
-## Usage
+Middleware runs _before_ cached content, so you can personalize static files and pages. Common examples of Middleware would be authentication, A/B testing, localized pages, bot protection, and more.
```

Comment:
Are localized pages something we recommend middleware for, rather than [i18n routing](https://nextjs.org/docs/advanced-features/i18n-routing)? If so, I think we're gonna need to make it clear in what situations each one is appropriate to avoid confusion. If not, we should cut this part of the sentence.

Reply:
Addressed with this [commit](https://github.com/vercel/next.js/pull/37992/commits/2f42c0a5fcc38e3125536df4fdff1206fae053a7)